### PR TITLE
Use a transaction when creating an artifact.

### PR DESCRIPTION
**Pull request by archseer**
Signed-off-by: Blaž Hrastnik blaz.hrast@gmail.com

### CODE REVIEW

File: artifact.go [hunk 1]
```diff
@@ -20,25 +20,34 @@ func NewArtifactRepo(db *DB) *ArtifactRepo {
...
 	if a.ID == "" {
 		a.ID = utils.UUID()
 	}
-	err := r.db.QueryRow("INSERT INTO artifacts (artifact_id, type, uri) VALUES ($1, $2, $3) RETURNING created_at",
+	err := tx.QueryRow("INSERT INTO artifacts (artifact_id, type, uri) VALUES ($1, $2, $3) RETURNING created_at",
 		a.ID, a.Type, a.URI).Scan(&a.CreatedAt)
 	if e, ok := err.(*pq.Error); ok && e.Code.Name() == "unique_violation" {
 		var deleted *time.Time
-		err = r.db.QueryRow("SELECT artifact_id, created_at, deleted_at FROM artifacts WHERE type = $1 AND uri = $2",
+		err = tx.QueryRow("SELECT artifact_id, created_at, deleted_at FROM artifacts WHERE type = $1 AND uri = $2",
```
Comment (titanous) [inline artifact.go]: I don't think reusing the `tx` here will work as it will have been automatically rolled back by the error.

Comment (archseer) [inline artifact.go]: Ah, that's something I wondered about. Why do we manually call `tx.Rollback()` on `err != nil` then?

Comment (titanous) [inline artifact.go]: `Rollback` also does some internal bookkeeping, so you should always call it in case of an error.

Comment (archseer) [inline artifact.go]: I could just use a transaction for the inner part -- queries that get executed when a formation of the same name already exists (but is deleted).

File: artifact.go [hunk 2]
```diff
@@ -20,25 +20,34 @@ func NewArtifactRepo(db *DB) *ArtifactRepo {
...
 		}
 		if deleted != nil {
-			err = r.db.Exec("UPDATE artifacts SET deleted_at = NULL WHERE artifact_id = $1", a.ID)
+			_, err = tx.Exec("UPDATE artifacts SET deleted_at = NULL WHERE artifact_id = $1", a.ID)
+			if err != nil {
+				tx.Rollback()
+				return err
+			}
 		}
+		tx.Commit()
```
Comment (titanous) [inline artifact.go]: Check the error returned by this.

File: artifact.go [hunk 3]
```diff
@@ -20,25 +20,37 @@ func NewArtifactRepo(db *DB) *ArtifactRepo {
...
+			}
+		}
+		err = tx.Commit()
+		if err != nil {
+			return err
 		}
 	}
 	a.ID = cleanUUID(a.ID)
-	return err
+	return nil
```
Comment (titanous) [inline artifact.go:54]: If `err` is not a `unique_violation` it isn't being checked now, you could add an `} else if err != nil {` to make things clearer.

File: artifact.go [hunk 4]
```diff
@@ -20,25 +20,37 @@ func NewArtifactRepo(db *DB) *ArtifactRepo {
...
+			_, err = tx.Exec("UPDATE artifacts SET deleted_at = NULL WHERE artifact_id = $1", a.ID)
+			if err != nil {
+				tx.Rollback()
+				return err
+			}
+		}
+		err = tx.Commit()
+		if err != nil {
+			return err
 		}
```
Comment (titanous) [inline artifact.go:49]: Use the condensed form here:

``` go
if err := tx.Commit(); err != nil {
    return err
}
```

File: artifact.go [hunk 5]
```diff
@@ -20,25 +20,38 @@ func NewArtifactRepo(db *DB) *ArtifactRepo {
...
 	err := r.db.QueryRow("INSERT INTO artifacts (artifact_id, type, uri) VALUES ($1, $2, $3) RETURNING created_at",
 		a.ID, a.Type, a.URI).Scan(&a.CreatedAt)
 	if e, ok := err.(*pq.Error); ok && e.Code.Name() == "unique_violation" {
+		tx, err := r.db.Begin()
+		if err != nil {
+			return err
+		}
 		var deleted *time.Time
-		err = r.db.QueryRow("SELECT artifact_id, created_at, deleted_at FROM artifacts WHERE type = $1 AND uri = $2",
+		err = tx.QueryRow("SELECT artifact_id, created_at, deleted_at FROM artifacts WHERE type = $1 AND uri = $2",
```
Comment (titanous) [inline artifact.go]: To take advantage of putting this in a transaction, I think we need to add `FOR UPDATE` to the end of the query.

**Comment by titanous** [comment]
LGTM pending last comment. Squash and then I'll merge.

**Comment by archseer** [comment]
Rebased.

**Pull request merged**